### DISCUSSION TITLE
Host controller ownership claim traps into endless loop

### DISCUSSION THREAD

**Discussion by davidgaosheng** [VendelinSlezak/BleskOS#21] [Q&A]
Hello,
I am learning EHCI driver source code from https://github.com/pdoane/osdev/blob/master/usb/ehci.c. During the initialization of ehci, the program bellow tries to claim the host controller ownership from BIOS and disable legacy support. The program got into endless loop waiting for  BIOS to release the ownership of the host controller and never return.   I think the program is of no difference with BleskOS except for that BleskOS does the same thing with only one command -- _pci_write(bus, device, function, ((mmio_ind(usb_controllers[usb_controllers_pointer].base+0x08)>>8) & 0xFF) , (1 << 24))_ and leaves; the former waits to ensure that an exchange of ownership of host controller be accomplishedin in a deterministic and reliable manner, as the Ehci Specification says.

 Please take some time to help me to identify the possible reason. Thank you so much!

#define USBLEGSUP                       0x00        // USB Legacy Support Extended Capability
#define UBSLEGCTLSTS                    0x04        // USB Legacy Support Control/Status

// ------------------------------------------------------------------------------------------------
// USB Legacy Support Register

#define USBLEGSUP_HC_OS                 0x01000000  // HC OS Owned Semaphore
#define USBLEGSUP_HC_BIOS               0x00010000  // HC BIOS Owned Semaphore

#define HCCPARAMS_EECP_MASK             (255 << 8)  // EHCI Extended Capabilities Pointer
#define HCCPARAMS_EECP_SHIFT            8

    //BleskOS:
    //pci_write(bus, device, function, ((mmio_ind(usb_controllers[usb_controllers_pointer].base+0x08)>>8) & 0xFF), (1 << 24)); //disable BIOS legacy support

    uint eecp = (hc->capRegs->hccParams & HCCPARAMS_EECP_MASK) >> HCCPARAMS_EECP_SHIFT;

    if (eecp >= 0x40)
    {
        // Disable BIOS legacy support
        uint legsup = PciRead32(id, eecp + USBLEGSUP);

        PciWrite32(id, eecp + USBLEGSUP, legsup | USBLEGSUP_HC_OS);

        if (legsup & USBLEGSUP_HC_BIOS)
        {
            PciWrite32(id, eecp + USBLEGSUP, legsup | USBLEGSUP_HC_OS);
            for (;;)
            {
                legsup = PciRead32(id, eecp + USBLEGSUP);
                if (~legsup & USBLEGSUP_HC_BIOS && legsup & USBLEGSUP_HC_OS)
                {
                    break;
                }
            }
        }
    }

**Comment by VendelinSlezak** [comment]
Well, my one line code for releasing BIOS ownership is horrible. Thanks for pointing to this, I will fix it.
Now to code, problem lies in line
``` 
for (;;)
```
this makes forever loop, if conditions in `if` are not met. Because this is case, method never returns. I would suggest to rewrite it to something like this:
```
       ticks = 0;
       while (ticks<TIMEOUT_VALUE)
        {
            legsup = PciRead32(id, eecp + USBLEGSUP);
            if (~legsup & USBLEGSUP_HC_BIOS && legsup & USBLEGSUP_HC_OS)
            {
                break;
            }
        }
        // error: BIOS ownership not released
```
where ticks is variabile incremented by some interrupt like PIT interrupt.

Anyway, where do you test this code? Emulator, real machine?

**Reply by davidgaosheng** [reply]
I tested this code in real machine, the same one on which I tested BleskOS.

as to the endless loop program, the endless problem is solved  if I arrange the initialization sequence as the following:
1. scan pci config space.
2. save each _UHCI_ host controller configuration information in _uhciconfig[ ]_ array
3. save each _EHCI_ host controller configuration information in _ehciconfig[ ]_ array.  
4. goto step 1 again till we have found all ehci and uhci host controller.
5. **first** intialize controllers in _ehciconfig[ ]_ array; and **then** the ones in _uhciconfig[]_ array.
What should be emphasized is that I initialized ehcis first, and then uhcis

the writer of the program initialized the host controller as  soon as he found it.

I don't understand why the modification I did to the program works.

**Reply by VendelinSlezak** [reply]
Then I would guess that problem is not in code you quoted, but somewhere else. For debugging purposes I would propose to consider these points:
- What happen if you at first initalize UHCI and then EHCI controllers
- What happen if you at first disable BIOS ownership of all UHCI and EHCI ports and only then process rest of initalization
- What port values are passed to initalization method if uhciconfig/ehciconfig arrays are not used
- Whether code really stucks in forever loop in code what you quote
- What is value in EHCI USBLEGSUP register after trying to release BIOS ownership in case if code stucks here.
…
